config/v1/types_ingress.go

```diff
@@ -41,13 +44,103 @@ type IngressSpec struct {
 	// certificate.
 	// +optional
 	AppsDomain string `json:"appsDomain,omitempty"`
+
+	// ComponentRoutes is a list of routes configured by a cluster-admin.
+	// The namespace and name of each componentRoute should match an existing Route.
+	//
+	// To determine the set of configurable Routes, look at namespace and name of entries in the
+	// .status.componentRoutes, where participating operators place current route status.
+	// +kubebuilder:validation:Optional
+	// +optional
+	ComponentRoutes []ComponentRouteSpec `json:"componentRoutes,omitempty"`
 }
 
 type IngressStatus struct {
+	// ComponentRoutes is where participating operators place the current route status for routes which the cluster-admin
+	// can customize hostnames and serving certificates.
+	//
+	// An operator that creates entries in this slice should clean them up during removal.
+	// An operator must also handle the case of deleted status without churn.
+	// +kubebuilder:validation:Optional
+	// +optional
+	ComponentRoutes []ComponentRouteStatus `json:"componentRoutes,omitempty"`
 }
 
-// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// ComponentRouteSpec allows for configuration of a route's hostname and serving certificate.
+type ComponentRouteSpec struct {
+	// Namespace is the namespace of the route to customize. It must be a real namespace. Using an actual namespace
+	// ensures that no two components will conflict and the same component can be installed multiple times.
+	// +kubebuilder:validation:Pattern=[a-z0-9]([-a-z0-9]*[a-z0-9])?
+	// +kubebuilder:validation:Required
+	// +required
+	Namespace string `json:"namespace"`
+
+	// Name is the logical name of the route to customize.
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+
+	// Hostname is the hostname that should be supported by the route.
+	// +kubebuilder:validation:Required
+	// +required
+	Hostname string `json:"hostname"`
```

Comment:
format hostname, compare https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/v1/types_jsonschema.go#L32

Reply:
you had mentioned to [format the hostname](https://github.com/openshift/api/pull/852#discussion_r585456366), the link mentioned that it should adhere to RFC 1034 but there is no example regex. I found an [example RFC 1034 compliant regex](https://regex101.com/library/SuU6Iq), but wanted to make sure that you were asking for the introduction of a pattern before moving forward.

Reply:
You can set the format property, no need for a regex. I guess it is `// +kubebuilder:validation:Format=hostname`.

Reply:
Ah I see - thank you!